include/picongpu/particles/debyeLength/Estimate.kernel

```diff
@@ -0,0 +1,228 @@
+                 * and so must be called once per supercell.
+                 *
+                 * @tparam T_Acc alpaka accelerator type
+                 * @tparam T_ElectronBox electron storage type
+                 * @tparam T_ResultBox result storage type, device box for Result
+                 *
+                 * @param acc alpaka accelerator
+                 * @param electronBox electron storage
+                 * @param supercellIdx supercell index, including guards
+                 * @param momentumVariance estimated variance in physical momentums
+                 * @param result device box of type Result with a single element
+                 *        to write kernel result to
+                 */
+                template<typename T_Acc, typename T_ElectronBox, typename T_EstimateBox>
+                DINLINE void check(
+                    T_Acc const& acc,
+                    T_ElectronBox const electronBox,
+                    pmacc::DataSpace<simDim> const supercellIdx,
+                    float_64 const sumWeighting,
+                    float3_64 const momentumVariance,
+                    T_EstimateBox estimateBox) const
+                {
+                    auto const supercellVolume
+                        = CELL_VOLUME * precisionCast<float_64>(SuperCellSize::toRT().productOfComponents());
+                    auto const electronDensity = sumWeighting / supercellVolume;
+
+                    /* For the Maxwell-Boltzmann distribution,
+                     * variance = mass * k * T, both for physical and macroparticle.
+                     * We assume this distribution and estimate k * T out of variance.
+                     * Applying the ideal gas law would result in the same estimate.
+                     */
+                    using Frame = typename T_ElectronBox::FrameType;
+                    auto const mass = static_cast<double>(picongpu::traits::frame::getMass<Frame>());
+                    auto const kT = (momentumVariance[0] + momentumVariance[1] + momentumVariance[2]) / (3.0 * mass);
+                    auto const temperatureKeV = kT * UNIT_ENERGY / UNITCONV_keV_to_Joule;
+                    auto const electronCharge = picongpu::traits::frame::getCharge<Frame>();
+                    auto const debyeLength = sqrt(EPS0 * kT / (electronDensity * electronCharge * electronCharge));
+
+                    cupla::atomicAdd(acc, &(estimateBox(0).sumWeighting), sumWeighting, ::alpaka::hierarchy::Blocks{});
+                    cupla::atomicAdd(
+                        acc,
+                        &(estimateBox(0).sumTemperatureKeV),
+                        sumWeighting * temperatureKeV,
+                        ::alpaka::hierarchy::Blocks{});
+                    cupla::atomicAdd(
+                        acc,
+                        &(estimateBox(0).sumDebyeLength),
+                        sumWeighting * debyeLength,
```

Comment:
You are computing the average Debye length, which is great.
But assuming we have a inhomogeneous plasma, would computing the minimal Debye length not reveal issues with resolutions much clearer, or are there numerical issues for the per-supercell evaluation?

Reply:
That's a good point, i should have at least added to tge output it's average. With min I was a bit confused how to do it properly. I think just taking min is susceptible to one unfortunately sampled supercell raising a false alarm. So it has to be some more sophisticated check i guess

Reply:
You could preform the check for resolving the Debye length already on every super cell and the just sum the Boolean result (fulfilled = 0, not fulfilled = 1). If the sum exceeds a "magic" threshold you could throw a warning.

Reply:
I agree that would be a more thorough check.

I actually tried to do almost what you are suggesting with the original version of this PR. I think it gets out of hand quite quickly, as then one needs to count "valid" supercells with at least 2 macroparticles and what ratio of those satisfy the check. And we need to somewhat arbitrarily decide what passes or not, see review of Klaus above, it probably should also depend on the particle shape. And it gets hardet to interpret.

I think i can add back this kind of check in addition to the average length output now.

Reply:
I think adding this would be very helpful.  But you are right, with only 2 ppc this might get out of hand due to numerical issues quickly.

Reply:
Yes, so I will try, but just to give my opinion since I've already considered it in the past. There are at least two fhings that could be problematic and that pushed me to do the implementation based on averages.

First, I assume that the initial momentum is a sum of a non-random part (like an initial drift) and a random part fully caused by the initial temperature. Since we need to fit distribution parameters (that are directly connected to temperature) of that random part based on the given sample, the sample needs to have a sufficient size. And so the larger volume considered, the better for this part.

However, while inside the simulation I also don't know how to split between the random and non-random parts. So I just assume the non-random part is slowly varying inside each sample volume size, and then I just use sample variance and assume it's fully caused by temperature. So that sample volume should be not too small and not too large, hence a supercell seems reasonable, also convenient computationally.

The averagizing done currently really helps to smooth out everything. I am thinking of adding another output in additing to that, saying something like "X% of supercells have < 1 cell per Debye length, Y% have < 2 cells, Z% have < 3 cells"

Reply:
Actually we are only interested in the Debye Length at highest density. If the Debye length is resolved in regions of highest density, it will be resolved everywhere. I think, this is the origin of @PrometheusPi comment and is the reason why a global average is only of limited use.

Is it possible to calculate the average Debye length only for the highest density, or within a ten percent variation of this? The error on the Debye length will then become 5%, which is negligible for resolution considerations. But the value of the Debye length we obtain is the one we need to judge on resolution.

Reply:
I suggest discussing this tomorrow in the developer meeting and summarise it here. I will reserve 10min for this topic.

Reply:
Agree with @psychocoderHPC . However as a question, isn't it possible that initial temperature is not constant in the sumulation domain? If so, it seems only density is not sufficient to determine the "interesting" region for the check.

Reply:
I think what you suggest @steindev is not easy to do, but better discuss tomorrow indeed.

Reply:
Actually, we could use the ´BASE_DENSITY´ for this estimate. This would be easy, wouldn't it? And as the typical workflow is to set `BASE_DENSITY` to the highest expected density, this would be sufficient. We could even compare the density you calculate to the `BASE_DENSITY` and, first, give a warning if a measured density value is larger than the `BASE_DENSITY`, second, calculate the Debye length only if the measured density is within a 10% variation (or so) around the `BASE_DENSITY` and calculate the average Debye length only from this reduced set of Debye lengths.

Reply:
Not sure if I follow. `BASE_DENSITY` is merely a unit, we do not require actual density to be close or not more than `BASE_DENSITY`. And a user doesn't have to initialize their macroparticles with our standard generators anyways. That's actually the whole issue with estimating Debye length: since we allow anything, there is nothing to rely upon before all the init pipeline is done and we have a set of macroparticles.

Reply:
`BASE_DENSITY` can not be used, as @sbastrakov said this is a unit. It is allowed to initialize a multiple of this value.

Reply:
Updated PR description with the points from VC discussion